src/main/scala/monocle/Optic.scala

```diff
@@ -0,0 +1,303 @@
+
+object Optic:
+  def id[A]: Optic[Get & ReverseGet, A, A] = 
+    POptic.id
+
+  object thatCan extends OpticConstructors(POptic.thatCan)
+
+end Optic
+
+
+object POptic:
+  def id[A, B]: POptic[Get & ReverseGet, A, B, A, B] = 
+    thatCan.convertBetween[A, B, A, B](identity)(identity)
+
+  object thatCan extends POpticConstructors
+
+end POptic
+
+final class POptic[+ThisCan, -S, +T, +A, -B] private[monocle](protected[monocle] val impl: OpticImpl[ThisCan, S, T, A, B]):
+
+  def andThen[ThatCan >: ThisCan, C, D](o: POptic[ThatCan, A, B, C, D]): POptic[ThatCan, S, T, C, D] =
+    POptic(impl.andThen(o.impl))
+
+  def index[I, S1 <: S, A1 >: A, V](i: I)
+      (using evIndex: Index[A1, I, V])
+      (using T <:< S1, A1 <:< B): Optic[ThisCan | (GetOption & Modify), S1, V] = 
+    asInstanceOf[Optic[ThisCan, S1, A1]].andThen(evIndex.index(i))
+
+  override def toString() = 
+    s"POptic(${impl})"
+
+end POptic
+
+extension [S, T, A, B] (self: POptic[GetMany, S, T, A, B])
+  def all(p: A => Boolean): S => Boolean =
+    s => self.impl.safe.foldMap(p)(s)(using Monoids.all)
+
+extension [S, A] (self: Optic[Get, S, A])
+  def choice[S1](other: Optic[Get, S1, A]): Optic[Get, Either[S, S1], A] =
+    Optic.thatCan.get(_.fold(self.get, other.get))
+
+extension [ThisCan, S, A] (self: Optic[ThisCan, S, A])
+  def each[C](using evEach: Each[A, C]): Optic[ThisCan | (GetMany & Modify), S, C] = 
+    self.andThen(evEach.each)
+
+extension [S, T, A, B] (self: POptic[GetMany, S, T, A, B])
+  def exist(p: A => Boolean): S => Boolean =
+    s => self.impl.safe.foldMap(p)(s)(using Monoids.any)
+
+extension [S, T, A, B] (self: POptic[GetMany, S, T, A, B])
+  def find(p: A => Boolean): S => Option[A] =
+    s => self.impl.safe.foldMap(a => Some(a).filter(p))(s)(using Monoids.firstOption)
+
+extension [S, T, A, B] (self: POptic[GetMany, S, T, A, B])
+  def fold(s: S)(using Monoid[A]): A =
+    self.impl.safe.foldMap(identity)(s)
+
+extension [S, T, A, B] (self: POptic[GetMany, S, T, A, B])
+  def foldMap[M: Monoid](f: A => M)(s: S): M = 
+    self.impl.safe.foldMap(f)(s)
+
+extension [S, T, A, B] (self: POptic[Get, S, T, A, B])
+  def get(s: S): A = 
+    self.impl.safe.get(s)
+
+extension [S, T, A, B] (self: POptic[GetMany, S, T, A, B])
+  def getAll(s: S): List[A] = 
+    self.impl.safe.foldMap[List[A]](List(_))(s)
+
+extension [S, T, A, B] (self: POptic[GetOneOrMore, S, T, A, B])
+  def getOneOrMore(s: S): NonEmptyList[A] = 
+    self.impl.safe.nonEmptyFoldMap[NonEmptyList[A]](NonEmptyList(_, Nil))(s)
+
+extension [S, T, A, B] (self: POptic[GetOption, S, T, A, B])
+  def getOption(s: S): Option[A] = 
+    self.impl.safe.getOption(s)
+
+extension [S, T, A, B] (self: POptic[GetOption & Modify, S, T, A, B])
+  def getOrModify(s: S): Either[T, A] = 
+    self.impl.safe.getOrModify(s)
+
+extension [S, T, A, B] (self: POptic[GetMany, S, T, A, B])
+  def headOption(s: S): Option[A] =
+    self.impl.safe.foldMap(Option.apply)(s)(using Monoids.firstOption)
```

Comment:
what is the rule for the definition of a function? When do we define it in `Optic` or in the `Impl`.